next/components/verify/useVerifyEid.tsx

```diff
@@ -0,0 +1,126 @@
+import { cityAccountClient } from '@clients/city-account'
+import { useMutation } from '@tanstack/react-query'
+import { useSsrAuth } from 'frontend/hooks/useSsrAuth'
+import {
+  FORM_SEND_EID_TOKEN_QUERY_KEY,
+  popSendEidMetadataVerify,
+  setVerifyEidMetadata,
+} from 'frontend/utils/metadataStorage'
+import { useRouter } from 'next/router'
+import { useRef, useState } from 'react'
+import { useEffectOnce } from 'usehooks-ts'
+
+import { environment } from '../../environment'
+import { useRefreshServerSideProps } from '../../frontend/hooks/useRefreshServerSideProps'
+import { ErrorWithName } from '../../frontend/utils/errors'
+
+export enum VerificationStatus {
+  INIT = 'INIT',
+  REDIRECTING = 'REDIRECTING',
+  VERIFYING = 'VERIFYING',
+  ERROR = 'ERROR',
+}
+
+/**
+ * This hook controls verifying identity with eID.
+ * Inspired by `useFormSend` hook.
+ *
+ * Verify with eID:
+ * 1. The user clicks on the "Overiť s eID" button.
+ * 2. `loginWithEid` is called. This triggers redirects to the eID login page.
+ *    Before redirecting, we store the metadata (that login was triggered as attempt to verify identity) in
+ *    the sessionStorage to be able to restore it after the user comes back.
+ * 3. If user successfully logs in, he/she is redirected to /nases/login with the token in the URL. As it is not possible
+ *    to parametrize the redirect URL, the login page retrieves the metadata from the sessionStorage and redirects
+ *    back to the verification page URL. The URL contains the token, which, if it's detected is immediately removed
+ *    from the URL and the user identity is verified using `verifyWithEid`.
+ */
+
+export const useVerifyEid = () => {
+  const router = useRouter()
+
+  const [verificationStatus, setVerificationStatus] = useState<VerificationStatus>(
+    // When the page is opened after redirect from slovensko.sk, we show VERIFYING state and call `verifyWithEid()`
+    router.query[FORM_SEND_EID_TOKEN_QUERY_KEY]
+      ? VerificationStatus.VERIFYING
+      : VerificationStatus.INIT,
+  )
+  const [verificationError, setVerificationError] = useState<Error | null>(null)
+
+  // As the token is immediately removed from the URL, we need to store it in a ref.
+  const sendEidTokenRef = useRef<string | null>(null)
+  const { tierStatus } = useSsrAuth()
+
+  const { refreshData } = useRefreshServerSideProps(tierStatus)
+
+  const { mutate: verifyWithEid, isPending: verifyWithEidIsPending } = useMutation({
+    mutationFn: () => {
+      return cityAccountClient.verificationControllerVerifyWithEid(
+        {
+          oboToken: sendEidTokenRef.current as string,
```

Comment:
Instead of type casting, we should check if `sendEidTokenRef.current` is present before calling the endpoint. (Even though it will probably never happen)